_posts/2019-12-06-integration-testing-middleware.adoc

```diff
@@ -0,0 +1,695 @@
+
+image::integration-testing-middleware/studio-intellij-switch-project-view.png[align="center"]
+
+
+The test execution via CUBA Studio is done through creating a run configuration for the test case and executing that run configuration. A test case or class can either be executed by the shortcut `CRTL+F5` (Windows) / `CMD+R` (Mac) or alternatively by using the Run icon on the test case / test class:
+
+image::integration-testing-middleware/running-test-via-cuba-studio.png[align="center"]
+
+CUBA Studio leverages the standard IntelliJ IDEA capabilities around testing and test execution. More information can be found on the https://www.jetbrains.com/help/idea/performing-tests.html[IDEA documentation].
+
+=== Automatic Visit Creation
+
+
+The first test case is about the automatic visit creation functionality of the Petclinic application. In the Visit browse screen, the user can create a new visit by entering the Pet Identification Number as a quick action. The software will go ahead and create a visit for the entered Pet.
+
+++++
+<a href="https://raw.githubusercontent.com/cuba-guides/cuba-petclinic-intro-to-datamanager/master/img/create-visit-for-pet.png"><img src="https://raw.githubusercontent.com/cuba-guides/cuba-petclinic-intro-to-datamanager/master/img/create-visit-for-pet.png"/></a>
+++++
+
+Part of the implementation of the feature is a CUBA service called `VisitService` that is triggered from the UI. It contains one method `Visit createVisitForToday(String identificationNumber)` that is the subject of the test case.
+
+WARNING: Although the feature includes a UI part, this will not be part of the automated testing. It is quite common to test certain sub parts of the implementation in isolation (as you will see). That being said, there is still a need for a real end-to-end test case to verify all the parts work together correctly. But this is out of scope for this guide.
+
+
+==== Implementation of the Functionality
+
+Before looking into the first test case, here is a recap of what the `VisitService` is doing in the implementation:
+
+1. it looks up the Pet by the given Identification Number
+a. in case the Pet is found it will use it
+b. in case no Pet is found, it will not create a Visit
+2. it creates a new instance of a Visit with the correct attributes
+a. it figures out the correct Pet and assigns it to the Visit
+b. it uses `today` as the `visitDate`
+3. it stores the newly created Visit
+
+The implementation of this functionality is shown below:
+
+.VisitServiceBean.java
+[source, java]
+----
+@Service(VisitService.NAME)
+public class VisitServiceBean implements VisitService {
+  @Inject
+  protected DataManager dataManager;
+  @Inject
+  protected TimeSource timeSource;
+  //...
+  @Override
+  public Visit createVisitForToday(String identificationNumber) {
+    return loadPetByIdentificationNumber(identificationNumber) // <1>
```

Comment:
The code differs from the app code in https://github.com/cuba-guides/cuba-petclinic-testing-middleware.
And don't you think it is excessively functional here? (just to my taste)

Reply:
> The code differs from the app code in https://github.com/cuba-guides/cuba-petclinic-testing-middleware.

Right. Main part was there before I decided to create a dedicated unit testing guide. I removed the code now

> And don't you think it is excessively functional here? (just to my taste)

Yes :) in fact I thought about that as well. Perhaps `.map().map()` is a little over the top. It is more my personal taste. Perhaps it is not the best idea to do it like that when educating people on their first unit test...

Reply:
I think I will change it 👍